src/partial_quicksync.rs

```diff
@@ -0,0 +1,203 @@
+  if !resp.status().is_success() {
+    anyhow::bail!(
+      "Failed to download file {}: HTTP status {}",
+      url,
+      resp.status()
+    );
+  }
+  let mut file = File::create(target_path).context("Failed to create file")?;
+  resp
+    .copy_to(&mut file)
+    .context("Failed to copy response to file")?;
+  Ok(())
+}
+
+fn decompress_file(input_path: &str, output_path: &str) -> Result<()> {
+  let input_file = File::open(input_path).context("Failed to open input file")?;
+  let output_file = File::create(output_path).context("Failed to create output file")?;
+
+  let mut reader = BufReader::new(input_file);
+  let mut writer = BufWriter::new(output_file);
+
+  let mut decoder = Decoder::new(&mut reader).context("Failed to create decoder")?;
+  decoder
+    .window_log_max(31)
+    .context("Failed to set window log max")?;
+
+  io::copy(&mut decoder, &mut writer).context("Failed to decompress")?;
+
+  Ok(())
+}
+
+fn execute_restore(target_conn: &Connection, restore_string: &str) -> Result<()> {
+  target_conn
+    .execute_batch(restore_string)
+    .context("Failed to execute restore")
+}
+
+pub fn partial_restore(layer_from: i64, target_db_path: &str, jump_back: usize) -> Result<()> {
+  let client = Client::new();
+  let base_url = get_base_url();
+  let user_version = get_user_version(target_db_path)?;
+  let remote_metadata = client
+    .get(format!("{}/{}/metadata.csv", base_url, user_version))
+    .send()?
+    .text()?;
+  let restore_string = client
+    .get(format!("{}/{}/restore.sql", base_url, user_version))
+    .send()?
+    .text()?;
+  let layer_from = if layer_from == 0 {
+    get_latest_from_db(target_db_path)?
+  } else {
+    layer_from
+  };
+  let start_points = find_start_points(layer_from, &remote_metadata, jump_back);
+  if start_points.is_empty() {
+    anyhow::bail!("No suitable restore point found");
+  }
+  println!("Found {} potential restore points", start_points.len());
+
+  for point in start_points {
+    let conn = Connection::open(target_db_path)?;
+    let previous_hash = get_previous_hash(point.from, &conn)?;
+
+    if previous_hash == &point.hash[..4] {
+      let source_db_path_zst = "backup_source.db.zst";
+      let source_db_path = "backup_source.db";
+
+      if download_file(
+        &client,
+        user_version,
+        point.from,
+        point.to,
+        &point.hash,
+        source_db_path_zst,
+      )
+      .is_err()
+      {
+        download_file(
+          &client,
+          user_version,
+          point.from,
+          point.to,
+          &point.hash,
+          source_db_path,
+        )?;
+      } else {
+        decompress_file(source_db_path_zst, source_db_path)?;
+      }
+
+      println!("Restoring from {} to {}...", point.from, point.to);
+      let start = Instant::now();
+      execute_restore(&conn, &restore_string)?;
+      fs::remove_file(source_db_path)?;
+      let _ = fs::remove_file(source_db_path_zst);
```

Comment:
How about moving the removing of .db.zst to after it is decompressed to execute it only when needed?
```suggestion
      } else {
        decompress_file(source_db_path_zst, source_db_path)?;
        fs::remove_file(source_db_path_zst)?;
      }

      println!("Restoring from {} to {}...", point.from, point.to);
      let start = Instant::now();
      execute_restore(&conn, &restore_string)?;
      fs::remove_file(source_db_path)?;
```

Reply:
True.